go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go

```diff
@@ -0,0 +1,243 @@
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("SetReadWrite", tablet.Alias)
+	require.NoError(t, err)
+
+	// wait for the vtgate to finish updating the new primary tablet
+	time.Sleep(1 * time.Second)
+
+	// verify that the vtgate will recognize the new primary tablet
+	qr, _ := vtgateConn.ExecuteFetch("show vitess_tablets", 100, true)
+	require.Equal(t, 3, len(qr.Rows), "wrong number of tablet records in healthcheck, expected %d but had %d", 3, len(qr.Rows))
+	require.Equal(t, "-80", qr.Rows[0][2].ToString())
+	require.Equal(t, "PRIMARY", qr.Rows[0][3].ToString())
+	require.Equal(t, "SERVING", qr.Rows[0][4].ToString())
+	require.Equal(t, tabletAlias(reparentTabletUID), qr.Rows[0][5].ToString())
+
+	// delete the old primary tablet
+	// This will ensure that the vtgate will experience the primary connection error if the switch would have to wait for the `tabletRefreshInterval`.
+	deleteTablet(t, clusterInstance.Keyspaces[0].Shards[0].Vttablets[0])
+
+	// verify that the vtgate will route the `INSERT` statement to the new primary tablet instead of the deleted tablet
+	qr, err = vtgateConn.ExecuteFetch("insert into customer(customer_id, email) values(2, 'dummy1')", 100, true) // -80
+	require.EqualValues(t, 1, qr.RowsAffected)
+	require.NoError(t, err)
+}
+
+func tabletAlias(tabletUID int) string {
+	return fmt.Sprintf("%s-%010d", cell, tabletUID)
+}
+
+func addTablet(t *testing.T, tabletUID int, tabletType string) *cluster.Vttablet {
+	tablet := &cluster.Vttablet{
+		TabletUID: tabletUID,
+		Type:      tabletType,
+		HTTPPort:  clusterInstance.GetAndReservePort(),
+		GrpcPort:  clusterInstance.GetAndReservePort(),
+		MySQLPort: clusterInstance.GetAndReservePort(),
+		Alias:     tabletAlias(tabletUID),
+	}
+	// Start Mysqlctl process
+	mysqlctlProcess, err := cluster.MysqlCtlProcessInstanceOptionalInit(tablet.TabletUID, tablet.MySQLPort, clusterInstance.TmpDirectory, !clusterInstance.ReusingVTDATAROOT)
+	require.Nil(t, err)
+	tablet.MysqlctlProcess = *mysqlctlProcess
+	proc, err := tablet.MysqlctlProcess.StartProcess()
+	require.Nil(t, err)
+
+	// Start vttablet process
+	tablet.VttabletProcess = cluster.VttabletProcessInstance(
+		tablet.HTTPPort,
+		tablet.GrpcPort,
+		tabletUID,
+		cell,
+		shards[0],
+		keyspaceName,
+		clusterInstance.VtctldProcess.Port,
+		tablet.Type,
+		clusterInstance.TopoProcess.Port,
+		clusterInstance.Hostname,
+		clusterInstance.TmpDirectory,
+		clusterInstance.VtTabletExtraArgs,
+		clusterInstance.DefaultCharset)
+
+	// wait for mysqld to be ready
+	err = proc.Wait()
+	require.Nil(t, err)
+
+	tablet.VttabletProcess.ServingStatus = ""
+	err = tablet.VttabletProcess.Setup()
+	require.Nil(t, err)
+
+	serving := tablet.VttabletProcess.WaitForStatus("SERVING", time.Duration(60*time.Second))
+	assert.Equal(t, serving, true, "Tablet did not become ready within a reasonable time")
+
+	t.Logf("Added tablet: %s", tablet.Alias)
+	return tablet
+}
+
+func deleteTablet(t *testing.T, tablet *cluster.Vttablet) {
```

Comment:
This also feels like something used throughout the tests and maybe we can avoid adding another copy.

Reply:
I have the same question here: https://github.com/vitessio/vitess/pull/12893#discussion_r1179908022.